pkg/channel/consolidated/reconciler/controller/consumer_group_watcher.go

```diff
@@ -0,0 +1,197 @@
+	Watch(watcherID string, callback ConsumerGroupHandler) error
+
+	// Forget removes all callbacks that correspond to the watcherID
+	Forget(watcherID string)
+
+	// List returns all the cached consumer groups that match matcher.
+	// It will return an empty slice if none matched or the cache is empty
+	List(matcher Matcher) []string
+}
+
+type WatcherImpl struct {
+	logger *zap.SugaredLogger
+	//TODO name?
+	watchers             map[string][]ConsumerGroupHandler
+	cachedConsumerGroups []string
+	adminClient          kafka.AdminClient
+	pollDuration         time.Duration
+	done                 chan struct{}
+}
+
+func NewConsumerGroupWatcher(ctx context.Context, ac kafka.AdminClient, pollDuration time.Duration) ConsumerGroupWatcher {
+	return &WatcherImpl{
+		logger:               logging.FromContext(ctx),
+		adminClient:          ac,
+		pollDuration:         pollDuration,
+		watchers:             make(map[string][]ConsumerGroupHandler),
+		cachedConsumerGroups: make([]string, 0),
+	}
+}
+
+func (w *WatcherImpl) Start() error {
+	w.logger.Infow("ConsumerGroupWatcher starting. Polling for consumer groups", zap.Duration("poll duration", w.pollDuration))
+	go func() {
+		for {
+			select {
+			case <-after(w.pollDuration):
+				// let's get current observed consumer groups
+				observedCGs, err := w.adminClient.ListConsumerGroups()
+				if err != nil {
+					w.logger.Errorw("error while listing consumer groups", zap.Error(err))
+					continue
+				}
+				var notify bool
+				var changedGroup string
+				// Look for observed CGs
+				for _, c := range observedCGs {
+					if !Find(w.cachedConsumerGroups, c) {
+						// This is the first appearance.
+						w.logger.Debugw("Consumer group observed. Caching.",
+							zap.String("consumer group", c))
+						changedGroup = c
+						notify = true
+						break
+					}
+				}
+				// Look for disappeared CGs
+				for _, c := range w.cachedConsumerGroups {
+					if !Find(observedCGs, c) {
+						// This CG was cached but it's no longer there.
+						w.logger.Debugw("Consumer group deleted.",
+							zap.String("consumer group", c))
+						changedGroup = c
+						notify = true
+						break
+					}
+				}
+				if notify {
+					cacheMtx.Lock()
+					w.cachedConsumerGroups = observedCGs
+					cacheMtx.Unlock()
+					w.notify(changedGroup)
+				}
+				done()
+			case <-w.done:
+				break
+			}
+		}
+	}()
+	return nil
+}
+
+func (w *WatcherImpl) Terminate() {
+	w.watchers = nil
+	w.cachedConsumerGroups = nil
+	w.done <- struct{}{}
+}
+
+// TODO explore returning a channel instead of a taking callback
+func (w *WatcherImpl) Watch(watcherID string, cb ConsumerGroupHandler) error {
+	w.logger.Debugw("Adding a new watcher", zap.String("watcherID", watcherID))
+	watchersMtx.Lock()
+	defer watchersMtx.Unlock()
+	cbs, ok := w.watchers[watcherID]
+	if !ok {
+		cbs = []ConsumerGroupHandler{}
+	}
+	w.watchers[watcherID] = append(cbs, cb)
+	// notify at least once to get the current state
+	cb()
+	return nil
+}
```

Comment:
IIUC, we call `Watch` every time we reconcile a channel. 
The registered callback will be called every 2 seconds if something changes, right?

Let's say we have the following events:
- Create channel (callback 1)
- Update channel (callback 2)
- then the loop in `Start` calls notify
- callback 1 and callback 2 are called, and both do the same thing, here: https://github.com/knative-sandbox/eventing-kafka/pull/182/files#diff-8d0207acc30e8507b21472d7c7ec7070441d085b47345fc60e6f53ec601da3fdR266-R311

Do we really need to register multiple callbacks for a given channel?

Reply:
nope.